### ISSUE TITLE
Interface method conflicts with normal method

### ISSUE THREAD

**Issue by Pauan** [fable-compiler/Fable#502]
### Description

I ran into this problem when I was trying to create a `View` abstract type:

```
open System

type TransactionId = uint32


[<AbstractClass>]
type View<'A>() =
    abstract member Subscribe : (TransactionId -> unit) -> IDisposable

    interface IObservable<'A> with
        member this.Subscribe out =
            this.Subscribe (fun _ -> ())


type Map<'A>(f, parent: View<'A>) =
    inherit View<'A>()

    override this.Subscribe fn =
        parent.Subscribe fn
```

I trimmed down the code a lot, so that it only shows the broken parts.

As you can see, `View` has a `Subscribe` method, but it also implements `IObservable` which also has a `Subscribe` method.

In F# this is not a problem, because interface methods cannot conflict with normal methods: you must disambiguate by using `:>`
### Expected behavior

```
export class View {
    constructor() {}

    Subscribe(out) {
        return this.Subscribe_0(_arg1 => {});
    }
}
declare(View);


export class Map extends View {
    constructor(f, parent) {
        super();
        this.parent = parent;
    }

    Subscribe_0(fn) {
        return this.parent.Subscribe_0(fn);
    }
}
declare(Map);
```
### Actual behavior

```
export class View {
    constructor() {}

    Subscribe(out) {
        return this.Subscribe_0(_arg1 => {});
    }
}
declare(View);


export class Map extends View {
    constructor(f, parent) {
        super();
        this.parent = parent;
    }

    Subscribe(fn) {
        return this.parent.Subscribe_0(fn);
    }
}
declare(Map);
```

`Map.prototype.Subscribe` should be `Map.prototype.Subscribe_0`
### Related information
- Fable Version `0.7.5-alpha.15`

**Comment by alfonsogarciacaro** [comment]
At the moment I've added a check that throws an error if interface method names conflict with type method names, but let's revisit this later to see if there's a better way to takle the issue
 :+1:

**Comment by Pauan** [comment]
@alfonsogarciacaro That's reasonable. Feel free to close this issue (or keep it open) as you wish.

**Comment by alfonsogarciacaro** [comment]
Latest alpha includes the `Mangle` attribute that can be added to interfaces to prevent conflicts with methods of the class or other interfaces. Can you please give it a try?

This is a different direction of the `NoMangle` proposal we discussed before. I know that having to use `Mangle` explicitly converts an F# feature into an opt-in in Fable. But after some consideration, I think this is a good way to keep good interoperability with JS while still having the possibility of mixing interface method names if needed. And it's also consistent with the `PassGenerics` attribute for example, which the developer must also use explicitly to get a default behaviour in F#.

**Comment by ncave** [comment]
@alfonsogarciacaro What about BCL interfaces that we cannot mangle? Are we out of luck there?

**Comment by alfonsogarciacaro** [comment]
@ncave I know this means the pattern in the example above is not possible in Fable (as it uses the `IObservable` interface from the BCL), but BCL interfaces are precisely one of the reasons that made be bet for the `Mangle` attribute. In my tests with `NoMangle` many lines in fable-core [like this one](https://github.com/fable-compiler/Fable/blob/88baf019ed21048dec3c85204f4ff2d9f747586d/src/fable/Fable.Core/ts/AsyncBuilder.ts#L146) were failing when calling methods like `Dispose`. Forcing the user to write the mangled name of the method would make interop in the opposite direction (JS->F#) more difficult, and it would also render [TS or Flow definitions of BCL interfaces](https://github.com/fable-compiler/Fable/blob/88baf019ed21048dec3c85204f4ff2d9f747586d/src/fable/Fable.Core/ts/Util.ts#L16-L18) useless.

**Comment by Pauan** [comment]
@alfonsogarciacaro What about [my solution](https://github.com/fable-compiler/Fable/pull/514#issuecomment-260556592)? It has good interop (both from JS->F# and F#->JS), it's compatible with TypeScript interfaces, and it has excellent performance.

The only downside is that it's still technically possible to have name conflicts, but it's extremely rare, so I consider it mostly a non-issue.

----

`IDisposable` would look like this in TypeScript:

```
interface IDisposable {
  System_IDisposable_Dispose(): void;
}
```

```
export class AsyncBuilder {
  Using<T extends IDisposable, U>(resource: T, binder: (x: T) => IAsync<U>) {
    return this.TryFinally(binder(resource), () => resource.System_IDisposable_Dispose());
  }
}
```

Not too bad at all, in my opinion. And TypeScript will catch it if you make a typo (or if you forget the `System_IDisposable_` prefix).

----

On a fundamental level, F# interfaces are not the same as TypeScript interfaces.

F# interfaces are *nominal*, TypeScript interfaces are *structural*.

F# interfaces cannot have method conflicts, TypeScript interfaces can.

I think the above solution is a good compromise: the prefixes make it a bit more cumbersome to use, but not by that much. And it preserves the benefit that F# interfaces cannot collide, while still being completely compatible with TypeScript interfaces.

**Comment by Pauan** [comment]
This has been fixed in `fable-compiler` version `0.7.5-alpha.37`! :tada:
